docs/keps/20200619-kubefed-pull-reconciliation.md

```diff
@@ -0,0 +1,314 @@
+* Improve the status property of federated resources to reflect the current state of
+the resources and not the propagation status only.
+
+## Proposal
+
+This approach splits the current kubefed architecture into two parts: agents (aka kubefed daemons) and a control-plane.
+It aims to decentralize the propagation and reconciliation towards the daemons.
+
+Users can choose the propagation and reconciliation model via a feature gate.
+With that, this approach provides backwards compatibility with the existing push model while enabling the users to choose this new model.
+
+In this pull-reconciler approach, we could highlight certain aspects that change in comparison with the pull-reconciler:
+
+* The kubefed agents are daemons running in the registered/joined kubefed clusters.
+* The control-plane logic changes to use a pull-based reconciliation and to rely on
+the kubefed daemons running on the target clusters handling most of the computation.
+* The management of federated resources remains in the cluster where the control-plane is deployed.
+The control-plane still handles the create, update, deletion of federated resources.
+
+<img src="./images/kubefedArch.jpg">
+
+The kubefed daemons are responsible of reconciling the cluster to the desired state
+and updating the status of the federated resources.
+
+This proposal is divided into two iterations in order to provide an experimental version of this new model as soon as possible:
+
+1) Initially, kubefed daemons have no restrictions to `watch`, `list` any federated resources and `write` the status of any federated resource.
+   Any daemon can view which resources are scheduled on any kubefed cluster, even if these federated resources are not placed in
+   its own cluster.
+   Kubefed daemons have to update the status of the kubefed cluster resources.
+   A periodically operation updates the health status of each kubefed cluster.
+
+2) The second iteration adds a shim API server to the kubefed control-plane components that sits between the Kubernetes API server and kubefed daemons.
+This solution provides a security middleware to control which resources can be viewed and accessed by each kubefed daemon.
+Additionally, this new daemon (the shim API server) increases the scalability of the control plane by reducing each daemon's impact on the Kubernetes API server.
+It watches events from the federated resources and does a fan-out to the kubefed daemon's.
+
+### Kubefed Daemon
+
+On each registered kubefed cluster, a daemon should be deployed to watch the state
+of the cluster and federated resources. Likewise this daemon should periodically
+reconcile the desired state to create/delete/update the federated resources.
+
+As done in the past, the `kubefedctl join` command creates a kubefed cluster in
+the control-plane cluster, and now deploys the kubefed daemon in the target cluster.
+This operation also exchanges the required permissions (tokens, kubeconfigs) and url to enable
```

Comment:
I would suggest we provide a way for users to configure the api server URL of the control plane cluster that'll be used by the kubefed daemon. In some case, the api server access might be exposed through an ingress gateway and the api server LB address might be a private LB address.

Reply:
Perhaps we could add an extra flag to this command to specify the api server url.